[package] Add dependency tracing function

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55167 [package] Add dependency tracing function**

**Summary**
This commit adds a function that uses `sys.setprofile` to trace the
execution of a callable in order to determine which modules it really
uses. The result of this trace can inform packaging decisions.

**Test Plan**
This commit adds a unit test to `test_analyze.py` that tests this
feature.

Differential Revision: [D27730805](https://our.internmc.facebook.com/intern/diff/D27730805)

Commit:
[WIP][package] Add tracing to exporter for automocking

Summary:
This commit uses `sys.setprofile` to trace the execution of an object
before it is packaging to automatically discover dependencies that are
*not* used and can be automatically mocked.

Test Plan:
This commit adds a unit test to `test_dependency_api.py` that tests this
feature.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "[package] Add tracing to exporter for automocking"


Summary:
This commit uses `sys.setprofile` to trace the execution of an object
before it is packaging to automatically discover dependencies that are
*not* used and can be automatically mocked.

Test Plan:
This commit adds a unit test to `test_dependency_api.py` that tests this
feature.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -334,7 +336,6 @@ def require_module(self, module_name: str, dependencies=True):
         and call `save_module` otherwise. Clients can subclass this object
         and override this method to provide other behavior, such as automatically mocking out a whole class
         of modules"""
-
```
Undo.

suo reviewed (changes requested):
I think we should move this out of the core packageexporter logic until we bake the functionality more fully. I think there should be a `torch/package/analyze` module or similar where we house some of this stuff (dependency visualization, pickler disassembler, trace mocker, etc.).

Commit:
Update on "[package] Add dependency tracing function"


**Summary**
This commit adds a function that uses `sys.setprofile` to trace the
execution of a callable in order to determine which modules it really
uses. The result of this trace can inform packaging decisions.

**Test Plan**
This commit adds a unit test to `test_analyze.py` that tests this
feature.

[ghstack-poisoned]

Commit:
Update on "[package] Add dependency tracing function"


**Summary**
This commit adds a function that uses `sys.setprofile` to trace the
execution of a callable in order to determine which modules it really
uses. The result of this trace can inform packaging decisions.

**Test Plan**
This commit adds a unit test to `test_analyze.py` that tests this
feature.

[ghstack-poisoned]

SplitInfinity commented:
Okay, I moved it to a separate directory as discussed, but now it only traces and returns dependencies. It won't be useful until we also land the change we discussed that adds a `get_dependency_graph` method to the `PackageExporter`, after which that graph can be combined with the results of this dependency tracing function to automatically mock dependencies.

suo commented on torch/package/analyze/trace_dependencies.py:
```diff
@@ -0,0 +1,56 @@
+            modules_used.add(module)
+
+    # Attach record_used_modules as the profiler function.
+    sys.setprofile(record_used_modules)
```
probably want to put this in a try-finally so that the profile doesn't get stuck on.

suo commented on test/package/test_analyze.py:
```diff
@@ -0,0 +1,27 @@
+    """Dependency analysis API tests."""
+
+    def test_trace_dependencies(self):
+        import automock
```
nit: this looks like a real package instead of test package, can we change the name 📦

Commit:
Update on "[package] Add dependency tracing function"


**Summary**
This commit adds a function that uses `sys.setprofile` to trace the
execution of a callable in order to determine which modules it really
uses. The result of this trace can inform packaging decisions.

**Test Plan**
This commit adds a unit test to `test_analyze.py` that tests this
feature.

[ghstack-poisoned]

Commit:
Update on "[package] Add dependency tracing function"


**Summary**
This commit adds a function that uses `sys.setprofile` to trace the
execution of a callable in order to determine which modules it really
uses. The result of this trace can inform packaging decisions.

**Test Plan**
This commit adds a unit test to `test_analyze.py` that tests this
feature.

Differential Revision: [D27730805](https://our.internmc.facebook.com/intern/diff/D27730805)

[ghstack-poisoned]
